core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -3474,15 +3482,18 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
   def foreach[R, E, Key, Key2, Value, Value2](
     map: Map[Key, Value]
   )(f: (Key, Value) => ZIO[R, E, (Key2, Value2)])(implicit trace: Trace): ZIO[R, E, Map[Key2, Value2]] =
-    if (map.isEmpty) ZIO.succeed(Map.empty)
-    else
-      ZIO.suspendSucceed {
-        val iterator = map.iterator
-        val builder  = Map.newBuilder[Key2, Value2]
-
-        val ft = f.tupled
-        ZIO.whileLoop(iterator.hasNext)(ft(iterator.next()))(builder += _).as(builder.result())
-      }
+    map.size match {
+      case 0 => ZIO.succeed(Map.empty)
+      case 1 => ZIO.suspendSucceed(f.tupled(map.head).map(Map.apply(_)))
+      case _ =>
+        ZIO.suspendSucceed {
+          val iterator = map.iterator
+          val builder  = Map.newBuilder[Key2, Value2]
+
+          val ft = f.tupled
```

Comment:
Is this better than unpacking the tuples manually?

Reply:
Just thought it makes the code cleaner. Changed to manual unpacking as it's more performant